tls/s2n_async_pkey.c

```diff
@@ -409,3 +422,134 @@ S2N_RESULT s2n_async_pkey_sign_free(struct s2n_async_pkey_op *op)
 
     return S2N_RESULT_OK;
 }
+
+int s2n_async_pkey_op_get_op_type(struct s2n_async_pkey_op *op, s2n_async_pkey_op_type *type)
+{
+    POSIX_ENSURE_REF(op);
+    POSIX_ENSURE_REF(type);
+
+    *type = op->type;
+
+    return S2N_SUCCESS;
+}
+
+int s2n_async_pkey_op_get_input_size(struct s2n_async_pkey_op *op, uint32_t *data_len)
+{
+    POSIX_ENSURE_REF(op);
+    POSIX_ENSURE_REF(data_len);
+
+    const struct s2n_async_pkey_op_actions *actions = NULL;
+    POSIX_GUARD_RESULT(s2n_async_get_actions(op->type, &actions));
+    POSIX_ENSURE_REF(actions);
+
+    POSIX_GUARD_RESULT(actions->get_input_size(op, data_len));
+
+    return S2N_SUCCESS;
+}
+
+static S2N_RESULT s2n_async_pkey_get_input_size_decrypt(struct s2n_async_pkey_op *op, uint32_t *data_len)
+{
+    struct s2n_async_pkey_decrypt_data *decrypt = &op->op.decrypt;
+    struct s2n_blob *in = &decrypt->encrypted;
+
+    *data_len = in->size;
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_async_pkey_get_input_size_sign(struct s2n_async_pkey_op *op, uint32_t *data_len)
+{
+    struct s2n_async_pkey_sign_data *sign = &op->op.sign;
+    struct s2n_hash_state *digest = &sign->digest;
+
+    uint8_t digest_length = 0;
+    RESULT_GUARD_POSIX(s2n_hash_digest_size(digest->alg, &digest_length));
+
+    *data_len = digest_length;
+
+    return S2N_RESULT_OK;
+}
+
+int s2n_async_pkey_op_get_input(struct s2n_async_pkey_op *op, uint8_t *data, uint32_t data_len)
+{
+    POSIX_ENSURE_REF(op);
+    POSIX_ENSURE_REF(data);
+
+    const struct s2n_async_pkey_op_actions *actions = NULL;
+    POSIX_GUARD_RESULT(s2n_async_get_actions(op->type, &actions));
+    POSIX_ENSURE_REF(actions);
+
+    POSIX_GUARD_RESULT(actions->get_input(op, data, data_len));
+
+    return S2N_SUCCESS;
+}
+
+static S2N_RESULT s2n_async_pkey_get_input_decrypt(struct s2n_async_pkey_op *op, uint8_t *data, uint32_t data_len)
+{
+    struct s2n_async_pkey_decrypt_data *decrypt = &op->op.decrypt;
+    struct s2n_blob *in = &decrypt->encrypted;
+
+    RESULT_ENSURE_LTE(in->size, data_len);
+
+    RESULT_CHECKED_MEMCPY(data, in->data, in->size);
+
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_async_pkey_get_input_sign(struct s2n_async_pkey_op *op, uint8_t *data, uint32_t data_len)
+{
+    struct s2n_async_pkey_sign_data *sign = &op->op.sign;
+
+    struct s2n_hash_state digest_copy = { 0 };
+    RESULT_GUARD_POSIX(s2n_hash_new(&digest_copy));
+    RESULT_GUARD_POSIX(s2n_hash_copy(&digest_copy, &sign->digest));
+
+    uint8_t digest_length = 0;
+    uint8_t digest_data[S2N_MAX_DIGEST_LEN] = { 0 };
+
+    RESULT_GUARD_POSIX(s2n_hash_digest_size(digest_copy.alg, &digest_length));
```

Comment:
As a safety check, do you want to verify that digest_length <= sizeof(digest_data) before you read the digest into digest_data?

Reply:
Yes that sounds good like a good idea, adding this